text/0000-code-literals.md

```diff
@@ -0,0 +1,268 @@
+- Feature Name: code_literals
+- Start Date: 2023-06-18
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a new kind of multi-line string literal for embedding code which
+plays nicely with `rustfmt`.
+
+# Motivation
+[motivation]: #motivation
+
+  - Embedding code as a literal string within a Rust program is often
+    necessary. A prominent example is the `sqlx` crate, which
+    has the user write SQL queries as string literals within the program.
+  - Rust already supports several kinds of multi-line string literal,
+    but none of them are well suited for embedding code.
+
+    1.  Normal string literals, eg. `"a string literal"`. These can be
+        written over multiple lines, but require special characters
+        to be escaped. Whitespace is significant within the literal,
+        which means that `rustfmt` cannot fix the indentation of the
+        code block. For example, beginning with this code:
+
+        ```rust
+        if some_condition {
+            do_something_with(
+                "
+                a nicely
+                indented code
+                string
+                "
+            );
+        }
+        ```
+
+        If the indentation is changed, such as by removing the
+        conditional, then `rustfmt` must re-format the code like so:
+
+        ```rust
+        do_something_with(
+            "
+                a nicely
+                indented code
+                string
+                "
+        );
+        ```
+
+        To do otherwise would be to change thange the value of
+        the string literal.
+    
+    2.  Normal string literals with backslash escaping, eg.
+        ```rust
+        "
+        this way\
+        whitespace at\
+        the beginning\
+        of lines can\
+        be ignored\
+        "
+        ```
+
+        This approach still suffers from the need to escape special
+        characters. The backslashes at the end of every line are
+        tedious to write, and are problematic if whitespace is
+        meaningful within the code. For example, if python code
+        was being embedded, then the indentation would be lost.
+        Finally, although `rustfmt` could in principle reformat
+        these strings, in practice doing so in a reasonable way
+        is complicated and so this has never been enabled.
```

Comment:
Do you think it would be more accurate to add a "by default" suffix here? I.e. we do have the still-unsable, off-by-default [format_strings](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#format_strings) option which can potentially touch such strings

```suggestion
        is complicated and so this has never been enabled by default.
```

Reply:
I've applied this change, although I will add that what `rustfmt` does implement doesn't necessarily align with what is reasonable for code. (eg. I believe it will do *more* than just fix indentation when this option is enabled, such as combining strings).

Reply:
Perhaps an alternative would then be to make the "reasonable" qualifier more explicitly targetted to the scope of strings with embedded code, whereas the example string this text is associated with doesn't have code and implies being more generally applicable to rustfmt behavior with strings (all these strings are of course currently all the same to rustfmt)

To be clear I'm not requesting another change, just sharing another thought around a different take in case you'd prefer something more along these lines over my prior proposal

e.g.

> Finally, although `rustfmt` could in principle reformat such strings, it doesn't do so by default and would be even more complicated to do so in a reasonable manner for strings with embedded code